### ISSUE TITLE
Allow specifying format of parameter in request

### ISSUE THREAD

**Issue by king-phyte** [SMILEY4/ktor-openapi-tools#168]
Hello. First of all, thank you for the excellent work you put into this well thought out and executed package.

I am using **ktor-swagger-ui:4.1.2**. I want a way to specify the format of a path parameter in the request section of the route documentation. See the sample below:

```kotlin
object UUIDSerializer : KSerializer<UUID> {
    override val descriptor = PrimitiveSerialDescriptor("UUID", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): UUID {
        return UUID.fromString(decoder.decodeString())
    }

    override fun serialize(encoder: Encoder, value: UUID) {
        encoder.encodeString(value.toString())
    }
}

// Using Type-safe routing (https://ktor.io/docs/server-resources.html)
@Resource("/users")
class Users() {
    @Resource("{id}") // <--- Trying to document this parameter
    class Id(
        val parent: Users = Users(),
        @Serializable(with = UUIDSerializer::class)
        @Schema(format = "uuid") // <--- This is not recognised in Type-safe routing
        @Format("uuid") // Not recognised
        val id: UUID,
    )
}

get<Users.Id>({
    request {
        pathParameter<UUID>("id") {
            description = "ID of the 
            format = "uuid" // <--- This is not available, but it is allowed in the OpenAPI specs
        }
    }
}) { user ->
    call.respond(status = HttpStatusCode.OK, message = mapOf("id" to user.id))
}
```

This generates the following (some fields are removed for brevity)

```json
"/users/{id}" : {
  "get" : {
    "parameters" : [ {
      "name" : "id",
      "in" : "path",
      "required" : true,
      "schema" : {
        "type" : "string",
        "title" : "String"
        "format": "uuid" // <--- this is missing
      }
    } ],
}
```

If you need any additional information, let me know and I will do my best to provide it. Thank you.

**Comment by ddenchev** [comment]
I am using a custom processor and would also love the ability to set the "format". 

```
schemas {
    generator = { type ->
        type
            .processKotlinxSerialization {
                customProcessor<Uuid> {
                    PrimitiveTypeData( ... )
                }    
```

i am pretty new to the library, but it feels ergonomic to be able to set the format through the custom processor override.

**Comment by king-phyte** [comment]
@ddenchev I was using a custom processor as well, but there was a deprecation somewhere in there. Did you hit it too? If you did, how did you proceed? For now, I am using type redirection so I need formatting to get help from Swagger

```kotlin
schemas {
            generator = { type ->
                type
                    .collectJacksonSubTypes({ it.processReflection() })
                    .processReflection {
                        redirect<OffsetDateTime, String>()
                        redirect<UUID, String>()
                    }
            }
        }
```

**Comment by ddenchev** [comment]
@king-phyte 
This is what my config currently looks like (I am using kotlinx serialization):

```
.processKotlinxSerialization {
	customProcessor<Uuid> {
		PrimitiveTypeData(
			id = TypeId(
				String::class.qualifiedName!!,
				emptyList(),
				"uuid"
			),
			simpleName = String::class.simpleName!!,
			qualifiedName = String::class.qualifiedName!!
		)
	}
}
```

As I said, I am new to the library so not sure about all the customization options and what rough edges are me not knowing about it.

The generated schema looks like this:
```

  "schema" : {
	"type" : "string",
	"title" : "String#uuid"
  },
```

In redoc, that looks like this:
<img width="673" alt="image" src="https://github.com/user-attachments/assets/815498db-cc90-422c-a138-6ec338fe6920" />

Instead of like this:
<img width="653" alt="image" src="https://github.com/user-attachments/assets/890d1d31-8949-47ff-8ce9-98f28f5076ab" />

**Comment by king-phyte** [comment]
@ddenchev Thank you for sharing that. It looks like we have hit the same wall.

**Comment by king-phyte** [comment]
@SMILEY4 Can you please take a look at this?

**Comment by SMILEY4** [comment]
Hi @king-phyte, @ddenchev,

I'm sorry for the confusion, i'm still trying to figure out how to make this simpler or document it better :)



For a bit of context:

Generating a schema for a kotlin type happens (very rougly) in two independent steps:

1.  collecting all the data for the type
2. generating a swagger schema from the collected data

The general idea for the uuids here would be:

- to use a custom processor for the uuid type to not use the default "data-collection" behavior in step 1 but to overwrite it with custom data where we can then specify anything about the type we want, i.e. a primitive type to be understood by step 2 as "string"
- step 2 (see `handleCoreAnnotations()`) already automatically looks at annotations like `@Format` and generates the swagger schema accordingly. The annotation can either come from manually adding it to fields or by specifying it as an annotation on the type itself in the custom processor from step 1. 



Assuming this model with the uuid field and the UUIDSerializer:

```kotlin
@Serializable
class Id(
    @Serializable(with = UUIDSerializer::class)
    val id: UUID, // no need for @Format("uuid") here. This can just be handled in the custom processor that is required anyway here.
)

object UUIDSerializer : KSerializer<UUID> {
    override val descriptor = PrimitiveSerialDescriptor("UUID", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): UUID {
        return UUID.fromString(decoder.decodeString())
    }

    override fun serialize(encoder: Encoder, value: UUID) {
        encoder.encodeString(value.toString())
    }
}
```



When using reflection, the schema generator config could look something like this (sticking close the the default plugin config here):    

```kotlin
install(SwaggerUI) {
    schemas {
        generator = { type ->
            type
            	.collectSubTypes()
                .processReflection {
                    // add a custom processor for type "UUID"
                    customProcessor<UUID> {
                        PrimitiveTypeData(
                            // needs a (unique) id for our type, overwriting UUID so taking that
                            id = TypeId.build(UUID::class.qualifiedName!!),
                            // qualified name must be the one of "String".
                            // The generator currently looks at the qualified name to
                            // determine the type of the swagger schema :/
                            qualifiedName = String::class.qualifiedName!!,
                            // simple name can be anything.
                            // Using uuid here since this will become
                            // the "title" in the swagger schema.
                            simpleName = UUID::class.simpleName!!,
                            annotations = mutableListOf(AnnotationData(
                                // adding the @Format annotation to our custom type here.
                                // This way it does not need to be added on every field in the models.
                                name = Format::class.qualifiedName!!,
                                values = mutableMapOf("format" to "uuid") // with value "uuid"
                            ))
                        )
                    }
                }
                .connectSubTypes()
                .handleNameAnnotation()
                .generateSwaggerSchema()
                 // @Format annotations will be handled in this step and is set in the swagger schema
                .handleCoreAnnotations()
                .withTitle(TitleType.SIMPLE)
                .compileReferencingRoot()
        }
    }
}
```

For kotlinx it's (almost) the same:

```kotlin
install(SwaggerUI) {
    schemas {
        generator = { type ->
            type
                .processKotlinxSerialization {
                    // register the custom processor the name matching the
                    // one defined in the custom KSerializer here:
                    //     PrimitiveSerialDescriptor("UUID", PrimitiveKind.STRING)
                    customProcessor("UUID") {
                        PrimitiveTypeData(
                            id = TypeId.build(UUID::class.qualifiedName!!),
                            qualifiedName = String::class.qualifiedName!!,
                            simpleName = UUID::class.simpleName!!,
                            annotations = mutableListOf(AnnotationData(
                                name = Format::class.qualifiedName!!,
                                values = mutableMapOf("format" to "uuid")
                            ))
                        )
                    }
                }
                .generateSwaggerSchema()
                 // set "format" in swagger schema according to @Format annotation
                .handleCoreAnnotations()
                .withTitle(SIMPLE)
                .compileReferencingRoot()
        }
    }
}
```



I hope this helps and clears up some of the confusion. If you have any open questions, please feel free to ask.

**Comment by king-phyte** [comment]
@SMILEY4 Thank you for the response.

It indeed works if I proceed with it like you described above. I was using custom processors by following the docs/examples, but as I said [in my comment above](https://github.com/SMILEY4/ktor-swagger-ui/issues/168#issuecomment-2564370206), I hit some deprecation warnings so I gave up and switched to redirection.
Perhaps you can update the examples with this.

Also, please evaluate the attached pull request, and decide if it is necessary or not (maybe for one-off moments of wanting to specify format), since the method you described here works fine if we always want to override the type.

Thank you once again!

**Comment by SMILEY4** [comment]
> I hit some deprecation warnings so I gave up and switched to redirection. Perhaps you can update the examples with this.

I updated the ktor-swagger-ui and schema-kenerator wiki and replaced mentions of the deprecated `withAutoTitle` with the updated `withTitle` functions. That should be the one you encountered ?

> Also, please evaluate the attached pull request, and decide if it is necessary or not

I don't think we need it since we have the custom processor that is required for larger modifications and the `@Format` annotation should imo suffice and work out of the box for the rest. I also would prefer to keep all the schema stuff in the schema-kenerator project.
Thank you for your pr/suggestion anyway!

**Comment by king-phyte** [comment]
> I updated the ktor-swagger-ui and schema-kenerator wiki and replaced mentions of the deprecated `withAutoTitle` with the updated `withTitle` functions. That should be the one you encountered ?

I don't remember which one exactly I encountered, but it was somewhere inside the PrimitiveTypeData

> I don't think we need it since we have the custom processor that is required for larger modifications and the `@Format` annotation should imo suffice and work out of the box for the rest. I also would prefer to keep all the schema stuff in the schema-kenerator project.
> 
> Thank you for your pr/suggestion anyway!

Alright then. You wouldn't mind if I closed this issue (and the PR), yes?

**Comment by SMILEY4** [comment]
> I don't remember which one exactly I encountered, but it was somewhere inside the PrimitiveTypeData

thanks, i'll have another look.

From my side this issue and the pr can be closed :+1:

**Comment by king-phyte** [comment]
Alright. Thank you yet once more for everything!

**Comment by king-phyte** [comment]
@SMILEY4 whilst I have you here, if you don't mind me asking, do you have plans of adding support for ReDoc?

**Comment by SMILEY4** [comment]
not until now, though i quickly looked into it and it seems quite simple to implement. 
The bigger problem might be how to add redoc to a project called ktor-**swagger**-ui :thinking:

**Comment by king-phyte** [comment]
Believe me when I say I have thought about that a lot 😅. But I figured having the functionalty first then worry about the naming later is a better approach

I considered forking the repo and maintaining a separate package for redoc, but I thought that would be a bit overboard, considering ReDoc consumes OpenAPI almost the same way as Swagger UI does and this package already does 99% of that work

What we need may be adding

```kotlin
install(SwaggerUI){
    redoc {...}
}
```

and

```kotlin
routing {
    route("docs/redoc") {
                redoc("/openapi.json")
            }
}
```

**Comment by SMILEY4** [comment]
:sweat_smile: regarding the naming, I'm currently trying to to split the project into an openapi-, swagger, (and possible future redoc) project (still keeping them in this single repo) and see if that would work - wanted to clean up a few things anyways :)

If that turns out to be more difficult or take longer, i could still add the redoc support like you suggested. I've basically got it working, might just make it a bit more configurable: https://github.com/SMILEY4/ktor-swagger-ui/compare/support-redoc?expand=1

**Comment by king-phyte** [comment]
Awesome stuff to be able to pull that in such a short time!

I agree that splitting it would be the best, since it could also mean cleaner and less coupled code and the ability to add other technologies (or custom ones) that consume the OpenAPI spec without much hassle. It may also make documenting the functionalites easier.
What do you say we create another issue to track the state and progress of that?

Also, if you need my help on anything, please don't hesitate to let me know. I will be glad to help.

Thank you!

**Comment by SMILEY4** [comment]
Thanks,
tracking this in another issue might be a good idea. Would you create one?

**Comment by king-phyte** [comment]
Gladly

**Comment by ddenchev** [comment]
> ```kotlin
>                 .handleCoreAnnotations()
> ```

Thanks for the explanation @SMILEY4! The annotations solution works perfectly! 

Regarding redoc support, it is easy enough to add it on the side, but would definitely use it if it comes included!
